File: backend/src/dogeapi/models/audit_log.py
"""Audit log model.

One row per mutating request. The middleware in :mod:`dogeapi.audit_log`
fills these in after every 2xx mutation; service modules can also call
``audit_log.record(...)`` directly for richer metadata.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from sqlalchemy import DateTime, Index, String, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column

from dogeapi.models.base import Base, UUIDPrimaryKeyMixin


class AuditLog(UUIDPrimaryKeyMixin, Base):
    """Audit log row.

    ``org_id`` and ``actor_id`` are *soft* references (no DB-level FK):
    the audit history must survive deletion of the underlying entity, and
    we never want a deletion cascade or SET NULL to mutate audit rows.
    """

    __tablename__ = "audit_logs"
    __table_args__ = (
        Index("ix_audit_logs_org_created", "org_id", "created_at"),
        Index("ix_audit_logs_org_action", "org_id", "action"),
    )

    org_id: Mapped[uuid.UUID | None] = mapped_column(
        PG_UUID(as_uuid=True),
        nullable=True,
        index=True,
    )
    actor_id: Mapped[uuid.UUID | None] = mapped_column(
        PG_UUID(as_uuid=True),
        nullable=True,
    )
    action: Mapped[str] = mapped_column(String(64), nullable=False)
    target_type: Mapped[str | None] = mapped_column(String(32), nullable=True)
    target_id: Mapped[str | None] = mapped_column(String(64), nullable=True)
    method: Mapped[str] = mapped_column(String(8), nullable=False)
    path: Mapped[str] = mapped_column(String(512), nullable=False)
    status_code: Mapped[int] = mapped_column(nullable=False)
    ip: Mapped[str | None] = mapped_column(String(64), nullable=True)
    user_agent: Mapped[str | None] = mapped_column(String(512), nullable=True)
    extra: Mapped[dict[str, Any]] = mapped_column(JSONB(), nullable=False, default=dict)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        index=True,
    )

File: frontend/src/lib/audit-log.ts
import { apiFetch } from "@/lib/api";

export type AuditLogEntry = {
  id: string;
  org_id: string | null;
  actor_id: string | null;
  action: string;
  target_type: string | null;
  target_id: string | null;
  method: string;
  path: string;
  status_code: number;
  ip: string | null;
  user_agent: string | null;
  extra: Record<string, unknown>;
  created_at: string;
};

export const listAuditLog = (
  slug: string,
  params: { action?: string; limit?: number; offset?: number } = {},
) => {
  const query = new URLSearchParams();
  if (params.action) query.set("action", params.action);
  if (params.limit !== undefined) query.set("limit", String(params.limit));
  if (params.offset !== undefined) query.set("offset", String(params.offset));
  const suffix = query.toString() ? `?${query.toString()}` : "";
  return apiFetch<AuditLogEntry[]>(`/orgs/${slug}/audit-log${suffix}`);
};
